Framework/Crystal/src/CalibrationHelpers.cpp

```diff
@@ -0,0 +1,165 @@
+
+  IComponent_const_sptr source = newInstrument->getSource();
+  IComponent_const_sptr sample = newInstrument->getSample();
+
+  V3D samplePos = detectorInfo.samplePosition();
+  if (samplePos != newSampPos) {
+    detectorInfo.setPosition(*sample, newSampPos);
+  }
+  V3D sourceRelPos = source->getRelativePos();
+  V3D sourcePos = detectorInfo.sourcePosition();
+  V3D parentSourcePos = sourcePos - sourceRelPos;
+  V3D source2sampleDir = samplePos - source->getPos();
+
+  double scalee = L0 / source2sampleDir.norm();
+  V3D newsourcePos = newSampPos - source2sampleDir * scalee;
+  V3D newsourceRelPos = newsourcePos - parentSourcePos;
+
+  detectorInfo.setPosition(*source, newsourceRelPos);
+}
+
+/**
+ * Updates the ParameterMap for newInstrument to reflect the changes in the
+ *associated panel information
+ *
+ * @param bankNames The names of the banks (panels) that will be updated
+ * @param newInstrument The instrument whose parameter map will be changed to
+ *reflect the new values below
+ * @param pos The quantity to be added to the current relative position, from
+ *old NewInstrument, of the banks in bankNames.
+ * @param rot The quantity to be added to the current relative rotations, from
+ *old NewInstrument, of the banks in bankNames.
+ * @param detWScale The factor to multiply the current detector width, from old
+ *NewInstrument, by to get the new detector width for the banks in bankNames.
+ * @param detHtScale The factor to multiply the current detector height, from
+ *old NewInstrument, by to get the new detector height for the banks in
+ *bankNames.
+ * @param rotCenters Rotate the centers of the panels(the same amount) with the
+ *rotation of panels around their center
+ * @param detectorInfo DetectorInfo object for the
+ */
+void CalibrationHelpers::fixUpBankPositionsAndSizes(
+    const std::vector<std::string> bankNames,
+    boost::shared_ptr<const Instrument> newInstrument, const V3D pos,
+    const Quat rot, double const detWScale, double const detHtScale,
+    bool rotCenters, DetectorInfo &detectorInfo) {
+  boost::shared_ptr<ParameterMap> pmap = newInstrument->getParameterMap();
+
+  for (const auto &bankName : bankNames) {
+    boost::shared_ptr<const IComponent> bank1 =
+        newInstrument->getComponentByName(bankName);
+    boost::shared_ptr<const Geometry::RectangularDetector> bank =
+        boost::dynamic_pointer_cast<const RectangularDetector>(bank1);
+
+    Quat RelRot = bank->getRelativeRot();
+    Quat newRelRot = rot * RelRot;
+    double rotx, roty, rotz;
+    quatToRotxRotyRotz(newRelRot, rotx, roty, rotz);
+
+    detectorInfo.setRotation(*bank, newRelRot);
+    //---------Rotate center of bank ----------------------
+    V3D center = bank->getPos();
+    V3D centerOrig(center);
+    if (rotCenters)
+      rot.rotate(center);
+
+    V3D pos1 = bank->getRelativePos();
+
+    detectorInfo.setPosition(*bank, pos + pos1 + center - centerOrig);
```

Comment:
I now notice that the `rotCenters` parameter is always `false`? Please check, and if this is true you can just remove the rotation part.